File: daemons/macos/src/Routing/RouteMatcher.swift
import Foundation

enum RouteMatcher {
    static func split(_ rawPath: String) -> (path: String, query: [String: String]) {
        if let markIndex = rawPath.firstIndex(of: "?") {
            return (
                String(rawPath[..<markIndex]),
                parseQuery(String(rawPath[rawPath.index(after: markIndex)...]))
            )
        }
        return (rawPath, [:])
    }

    static func match(_ path: String, pattern: String) -> [String: String]? {
        let pathParts = path.split(separator: "/", omittingEmptySubsequences: true)
        let patternParts = pattern.split(separator: "/", omittingEmptySubsequences: true)
        if pathParts.count == patternParts.count {
            var params: [String: String] = [:]
            for (p, pat) in zip(pathParts, patternParts) {
                if pat.hasPrefix(":") {
                    params[String(pat.dropFirst())] = String(p)
                } else if p != pat {
                    return nil
                }
            }
            return params
        }
        return nil
    }

    private static func parseQuery(_ raw: String) -> [String: String] {
        var result: [String: String] = [:]
        for pair in raw.split(separator: "&") {
            if let eq = pair.firstIndex(of: "=") {
                let key = String(pair[..<eq]).removingPercentEncoding ?? String(pair[..<eq])
                let valueRaw = pair[pair.index(after: eq)...]
                result[key] = String(valueRaw).removingPercentEncoding ?? String(valueRaw)
            }
        }
        return result
    }
}

File: daemons/linux/src/Routing/RouteMatcher.js
export function split(rawPath) {
  const question = rawPath.indexOf('?')
  if (question === -1) {
    return { path: rawPath, query: {} }
  }
  return {
    path: rawPath.slice(0, question),
    query: Object.fromEntries(new URLSearchParams(rawPath.slice(question + 1)))
  }
}

export function match(path, pattern) {
  const pathParts = path.split('/').filter(Boolean)
  const patternParts = pattern.split('/').filter(Boolean)
  if (pathParts.length !== patternParts.length) {
    return null
  }
  const params = {}
  for (let index = 0; index < pathParts.length; index += 1) {
    if (patternParts[index].startsWith(':')) {
      params[patternParts[index].slice(1)] = pathParts[index]
    } else if (pathParts[index] !== patternParts[index]) {
      return null
    }
  }
  return params
}
